presto-main/src/main/java/com/facebook/presto/dispatcher/FailedDispatchQueryFactory.java

```diff
@@ -14,50 +14,11 @@
 package com.facebook.presto.dispatcher;
 
 import com.facebook.presto.Session;
-import com.facebook.presto.event.QueryMonitor;
-import com.facebook.presto.execution.ExecutionFailureInfo;
-import com.facebook.presto.execution.LocationFactory;
-import com.facebook.presto.server.BasicQueryInfo;
 import com.facebook.presto.spi.resourceGroups.ResourceGroupId;
 
-import javax.inject.Inject;
-
 import java.util.Optional;
-import java.util.concurrent.ExecutorService;
-
-import static com.facebook.presto.util.Failures.toFailure;
-import static java.util.Objects.requireNonNull;
 
-public class FailedDispatchQueryFactory
+public interface FailedDispatchQueryFactory
```

Comment:
can you explain why you extract an interface here?

Reply:
also add javadoc please (even though it's not new)

Reply:
From high-level point of view, DispatchQuery is created from the DispatchQueryFactory, of which there can be two implementations, LocalDispatchQueryFactory which creates a LocalDispatchQuery, and a possible RemoteDispatchQueryFactory, which could create a RemoteDispatchQuery. Analogously, there could be two implementations to create a FailedDispatchQuery, and that is what this change does. 

From a code organization point of view, FailedLocalDispatchQueryFactory takes a dependency on QueryMonitor, which brings in many heavy weight classes, such as FunctionAndTypeManager, HistoryBasedPlanStatisticsTracker, and more, which in turn, bring in a laundry list of more dependencies. An alternative approach to extracting this interface, could be to split QueryMonitor into a DispatchQueryMonitor, which would have queryCreated and queryImmediateFailure methods, and a ExecutingQueryMonitor, which would have the rest of the methods (like queryCompletedEvent). This way, dependencies such as FunctionAndTypeManager, would not need to be wired into DispatchQueryMonitor. But this would be a much larger refactor, and I felt it cleaner to simply extract the interface from FailedDispatchQueryFactory itself, and allow for multiple implementations, especially since the underlying implementation is not complex, and there would not be much benefit from code re-use.

Reply:
Ultimately, the idea is to have other implementations, but the goal for this PR is not really about the specific interfaces, but about removing dependencies in queueing/dispatching code on classes not needed for queueing/dispatching, so it's fine if the interfaces need to change as the RFC is finalized. The end goal that this is working towards is that it should be possible to move queueing stuff out of Presto main to a separate package. There's certainly overlap between determining the right interfaces and figuring out how best to separate out dependencies, but since disentangling the dependencies is itself a substantial undertaking, and most of the work would be needed even if interfaces change, we want to get started with that process now.